### PR TITLE
Add TimeEffect effect for getting current time

### LINKED ISSUE #1502: Make "get current time" effect
There are several places where we have a `Has (Lift IO)` constraint *only* because we need to get the current time.  Instead, we should create a custom effect for getting the current time, and then interpret it into IO later.  This way we have a much clearer picture of what effects we actually need.

I'll probably get around to this sooner or later, but could also be a nice issue for someone who wants to learn more about how `fused-effects` works.

### PR DESCRIPTION
**Pull request by p3rsik**
fixes #1502

Add TimeEffect effect for getting current time.

### CODE REVIEW

**Comment by p3rsik** [comment]
@byorgey I'm not sure code became cleaner, as was your original intention, I assume, but now you know when you need to do something related to time, instead of just knowing that function deals with `IO` somehow. Perhaps extending the scope a bit and adding more effect for currently used IO actions would be a bit better. Or making another issue for that, idk.

File: src/Swarm/Util/Effect.hs [hunk 1]
```diff
@@ -78,3 +88,18 @@ forMW = flip traverseW
 
 simpleErrorHandle :: ThrowC SystemFailure IO a -> IO a
 simpleErrorHandle = either (fail . prettyString) pure <=< runThrow
+
+-- | Effect for things related to time
+data TimeEffect (m :: Type -> Type) k where
```
[the remark this answers is not in the document]
Comment (byorgey) [reply src/Swarm/Util/Effect.hs:93]: Putting this in `Swarm.Util.Effect` is fine for now, but longer term, if we want to start adding more custom effects, we won't want to stuff them all into this module.  What about putting it in a new module like `Swarm.Effect.Time`?  I'll leave it up to you whether you want to move it now (you'd have to add the new module to the list in the `.cabal` file) or wait and do that later, either is fine with me.

File: src/Swarm/Game/Step.hs [hunk 2]
```diff
@@ -1017,7 +1026,7 @@ addSeedBot e (minT, maxT) loc ts =
 -- | Interpret the execution (or evaluation) of a constant application
 --   to some values.
 execConst ::
-  (HasRobotStepState sig m, Has (Lift IO) sig m) =>
+  (HasRobotStepState sig m, Has TimeEffect sig m, Has (Lift IO) sig m) =>
```
Comment (byorgey) [inline src/Swarm/Game/Step.hs:1029]: I'm glad there are some cases (e.g. `doGrab`) where we were able to simply replace the `Lift IO` effect with `TimeEffect`.  I assume we still need the `Lift IO` in the other cases?

**Review by byorgey — APPROVED** [review]
Looks good to me, other than a few minor comments below.  I'll let you address those as you see fit (though happy to discuss further if you'd like) and then add the `merge me` label once you're happy with it.

**Pull request merged**